### PR TITLE
velero: Add provider field and remove default configuration in restic plugin

### PR DESCRIPTION
**Pull request by ipochi**
Brings back the addition of `provider` field in the component configuration.
This is done so to be able to provide default configurations for the plugins.

Also it removes the default BackupStorageLocation config in restic
plugin.

Name is optional and its default value is `default` in Helm chart.
Provider is required and making it a default of `aws` doesn't make sense
as the user has to provide  value to it anyway.

Region is also specifc to the provider used. If not `aws` then there is
no need to input that value.

Release notes: 

Add `provider` field in the component configuration.
example:

```
component `velero` {
  .
  .
  provider = "openebs"
 
  openebs {
    .
    .
  }
}
```

Signed-off-by: Imran Pochi <imran@kinvolk.io>

### CODE REVIEW

**Comment by knrt10** [comment]
Can you please address check linter and build errors?

**Comment by ipochi** [comment]
> Can you please address linter and build errors?

yep, I am looking into the bug introduced by this fix.

File: pkg/components/velero/component.go [hunk 1]
```diff
@@ -72,6 +72,9 @@ func newComponent() *component {
 			Enabled:        false,
 			ServiceMonitor: false,
 		},
+		Azure:   &azure.Configuration{},
```
Comment (invidian) [inline pkg/components/velero/component.go:75]: How is this related to default restic configuration removal?

File: pkg/components/velero/restic/restic.go [hunk 2]
```diff
@@ -44,11 +44,7 @@ type BackupStorageLocation struct {
 // NewConfiguration returns the default restic configuration.
 func NewConfiguration() *Configuration {
 	return &Configuration{
-		BackupStorageLocation: &BackupStorageLocation{
```
Comment (invidian) [inline pkg/components/velero/restic/restic.go:47]: Do we also need a documentation change for this?

Comment (ipochi) [reply pkg/components/velero/restic/restic.go:47]: I don't think so.

Comment (invidian) [reply pkg/components/velero/restic/restic.go:47]: All those 3 parameters are marked in documentation as `required = false`. I though this is because we were setting a default values for them. So they are actually optional?

**Review by invidian — COMMENTED** [review]
LGTM, except the docs for second commit (discussed in other comment).

File: pkg/components/velero/component.go [hunk 3]
```diff
@@ -189,15 +193,15 @@ func (c *component) getSupportedProviders() []string {
 func (c *component) getProvider() (provider, error) {
 	providers := []provider{}
 
-	if c.Azure != nil {
+	if c.Provider == "azure" {
```
Comment (invidian) [inline pkg/components/velero/component.go:196]: Actually, append logic doesn't make sense anymore if we can only have one provider configured at all times. I'd just replace entire function body with either switch/case or map lookup.

**Review by invidian — CHANGES_REQUESTED** [review]

File: pkg/components/velero/restic/restic.go (hunk 2 above)
Comment (ipochi) [reply pkg/components/velero/restic/restic.go:47]: updated the docs.

File: docs/configuration-reference/components/velero.md [hunk 4]
```diff
@@ -37,6 +37,7 @@ Velero component configuration example:
 # velero.lokocfg
 component "velero" {
 
+  #provider = "azure/openebs/restic"
```
Comment (knrt10) [inline docs/configuration-reference/components/velero.md:40]: I don't know but no space here seems a little bit weird.

Comment (invidian) [reply docs/configuration-reference/components/velero.md:40]: What space you mean? Blank line separating parameter from block below?

File: docs/configuration-reference/components/velero.md [hunk 5]
```diff
@@ -110,6 +111,7 @@ Table of all the arguments accepted by the component.
 | Argument                                             | Description                                                                                                                 | Default                                           | Type   | Required |
 |------------------------------------------------------|-----------------------------------------------------------------------------------------------------------------------------|---------------------------------------------------|--------|----------|
 | `namespace`                                          | Namespace to install Velero.                                                                                                | "velero"                                          | string | false    |
+| `provider`                                           | Provider sets which provider block to use for the configuration. Supported values `azure`, `openebs`, `restic`.             | -                                                 | string | true     |
```
Comment (invidian) [inline docs/configuration-reference/components/velero.md:114]:
```suggestion
| `provider`                                           | Provider sets which provider block to use for the configuration. Supported values `azure`, `openebs` and `restic`.             | -                                                 | string | true     |
```

File: docs/configuration-reference/components/velero.md [hunk 6]
```diff
@@ -145,10 +147,10 @@ Table of all the arguments accepted by the component.
...
-| `restic.backup_storage_location.name`                | Name for backup location object on the cluster.                                                                             | -                                                 | string | false    |
-| `restic.backup_storage_location.region`              | Cloud provider region for storing snapshots.                                                                                | `eu-west-1`                                       | string | false    |
+| `restic.backup_storage_location.name`                | Name for backup location object on the cluster.                                                                             | "default"                                         | string | false    |
+| `restic.backup_storage_location.region`              | Cloud provider region for storing snapshots. Required if `restic.backup_storage_location.provider = aws`                    | -                                                 | string | false    |
```
Comment (invidian) [inline docs/configuration-reference/components/velero.md:153]:
```suggestion
| `restic.backup_storage_location.region`              | Cloud provider region for storing snapshots. Required if `restic.backup_storage_location.provider = aws`.                   | -                                                 | string | false    |
```

**Review by invidian — CHANGES_REQUESTED** [review]
Some nits.

File: docs/configuration-reference/components/velero.md (hunk 4 above)
Comment (knrt10) [reply docs/configuration-reference/components/velero.md:40]: Yes, I meant `# provider` as we are using it in line with provider name in test file.

File: docs/configuration-reference/components/velero.md [hunk 7]
```diff
@@ -110,6 +111,7 @@ Table of all the arguments accepted by the component.
 | Argument                                             | Description                                                                                                                 | Default                                           | Type   | Required |
 |------------------------------------------------------|-----------------------------------------------------------------------------------------------------------------------------|---------------------------------------------------|--------|----------|
 | `namespace`                                          | Namespace to install Velero.                                                                                                | "velero"                                          | string | false    |
+| `provider`                                           | Provider sets which provider block to use for the configuration. Supported values `azure`, `openebs` and `restic`.          | -                                                 | string | true     |
```
Comment (invidian) [inline docs/configuration-reference/components/velero.md:114]:
```suggestion
| `provider`                                           | Provider sets which provider block to use for the configuration. Supported values are: `azure`, `openebs` and `restic`.          | -                                                 | string | true     |
```

**Review by invidian — DISMISSED** [review]
LGTM

**Review by invidian — APPROVED** [review]

**Review by knrt10 — APPROVED** [review]
